lib/winthrop-client-ruby/refresh_token.rb

```diff
@@ -0,0 +1,76 @@
+require 'uri'
+require 'net/http'
+require 'json'
+
+module WinthropClient
+  class RefreshToken
+    @token = nil
+    @expires_at = nil
+
+    TOKEN_GRANT_TYPE = 'client_credentials'.freeze
+    CONTENT_TYPE = 'application/x-www-form-urlencoded'.freeze
+    EXPIRY_IN_MINUTES = 60
+
+    class << self
+      attr_accessor :client_id, :client_secret, :host
+
+      def access_token
+        if @token.nil? || Time.now >= @expires_at
+          generate_access_token
+        else
+          @token
+        end
+      end
+
+      private
+
+      # Generates a new access token
+      def generate_access_token
+        http = build_http
+        request = build_request
+        response = http.request(request)
+        handle_response(response)
+      end
+
+      # Builds the HTTP connection
+      def build_http
+        uri = URI.parse(host)
+        http = Net::HTTP.new(uri.host, uri.port)
+        http.use_ssl = (uri.scheme == 'https')
+        http
+      end
+
+      # Builds the request for the token
+      def build_request
+        uri = URI.parse(host)
+        request = Net::HTTP::Post.new(uri)
+        request['Content-Type'] = CONTENT_TYPE
+        request.set_form_data(token_params)
+        request
+      end
+
+      # Returns the parameters required for the token request
+      def token_params
+        {
+          grant_type: TOKEN_GRANT_TYPE,
+          client_id: client_id,
+          client_secret: client_secret
+        }
+      end
+
+      # Handles the HTTP response and caches the token
+      def handle_response(response)
+        if response.is_a?(Net::HTTPSuccess)
+          parsed_response = JSON.parse(response.body)
+          @token = parsed_response['access_token']
+
+          # Set the expiry time to 1 hour from now
+          @expires_at = Time.now + (EXPIRY_IN_MINUTES * 60)
+          @token
+        else
+          raise "Failed to retrieve access token: #{response.body}"
```

Comment:
I'd recommend using [WinthropClient::ApiError](https://github.com/winthrop-intelligence/winthrop-client-ruby/blob/main/lib/winthrop-client-ruby/api_client.rb) for your error class here for consistency with the rest of the client.

Reply:
Done.